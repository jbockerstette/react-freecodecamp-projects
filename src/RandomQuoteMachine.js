import React from 'react';
import './RandomQuoteMachine.css';

const colors = [
  '#16a085',
  '#27ae60',
  '#2c3e50',
  '#f39c12',
  '#e74c3c',
  '#9b59b6',
  '#FB6964',
  '#342224',
  '#472E32',
  '#BDBB99',
  '#77B1A9',
  '#73A857'
];

export const Grid = ({ bgColor, ...rest }) => (
  <div className="grid" style={{ background: bgColor }} {...rest} />
);
export const Button = ({ bgColor, ...rest }) => (
  <button className="btn" style={{ background: bgColor }} {...rest} />
);
export const ButtonLink = ({ className, bgColor, ...rest }) => {
  const cls = ['btn', className].join(' ');
  return <a className={cls} style={{ background: bgColor }} {...rest} />;
};

export const Quote = ({ quote, color, opacity, ...rest }) => (
  <p {...rest} style={{ color, opacity }}>
    <i className="fa fa-quote-left" />
    {` ${quote} `}
    <i className="fa fa-quote-right" />
  </p>
);

export const Author = ({ author, color, opacity, ...rest }) => (
  <p {...rest} style={{ color, opacity }}>{`- ${author}`}</p>
);

class RandomQuoteMachine extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      quotes: [{ quote: '', author: '' }],
      quote: 0,
      colorIndex: 0,
      opacity: 1
    };
    this.handleNextQuote = this.handleNextQuote.bind(this);
  }

  async componentDidMount() {
    try {
      const json = await fetch(
        'https://gist.githubusercontent.com/camperbot/5a022b72e96c4c9585c32bf6a75f62d9/raw/e3c6895ce42069f0ee7e991229064f167fe8ccdc/quotes.json'
      );
      const quotes = await json.json();
      this.setState(() => ({ quotes: quotes.quotes }));
    } catch (error) {
      console.error(error);
    }
  }

  static getTwitterLink(q) {
    if (q) {
      return `https://twitter.com/intent/tweet?hashtags=quotes&related=freecodecamp&text='${encodeURIComponent(
        `"${q.quote}" ${q.author}`
      )}`;
    }
    return '';
  }

  handleNextQuote() {
    this.setState(() => ({ opacity: 0 }));
    setTimeout(() => {
      const quote = Math.round(Math.random() * this.state.quotes.length - 1);
      this.setState(({ colorIndex }) => {
        let nextColor = colorIndex + 1;
        if (nextColor === colors.length) {
          nextColor = 0;
        }
        return { quote, colorIndex: nextColor, opacity: 1 };
      });
    }, 1000);
  }

  render() {
    const { quotes, quote, colorIndex, opacity } = this.state;
    const q = quotes[quote];
    const color = colors[colorIndex];
    const href = RandomQuoteMachine.getTwitterLink(q);
    console.log(q);
    return q ? (
      <Grid bgColor={color}>
        <div id="quote-box">
          <Quote id="text" color={color} opacity={opacity} quote={q.quote} />
          <Author
            id="author"
            color={color}
            opacity={opacity}
            author={q.author}
          />
          <ButtonLink
            className="fa fa-twitter"
            id="tweet-quote"
            bgColor={color}
            href={href}
            target="_blank"
          />
          <Button id="new-quote" onClick={this.handleNextQuote} bgColor={color}>
            New Quote
          </Button>
        </div>
      </Grid>
    ) : null;
  }
}
export default RandomQuoteMachine;
