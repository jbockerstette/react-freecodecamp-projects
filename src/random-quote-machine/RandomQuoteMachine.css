#quote-box {
  display: grid;
  grid-template-columns: 1fr 1fr;
  width: 450px;
  border: solid whitesmoke 1px;
  border-radius: 5px;
  padding: 32px;
  background: whitesmoke;
}

#quote-box #text {
  grid-column: 1 / -1;
  font-size: 24px;
  text-align: start;
  transition: all 1.5s ease;
  margin: 0;
}

#quote-box #author {
  grid-column: 1 / -1;
  justify-self: end;
  transition: all 1.5s ease;
}

#quote-box #new-quote {
  grid-column: 2 / -1;
  justify-self: end;
}

#quote-box #tweet-quote {
  grid-column: 1 / 2;
  justify-self: start;
}

.btn {
  font-size: normal;
  border: 0;
  color: whitesmoke;
  padding: 12px;
  border-radius: 5px;
  outline: none;
  cursor: pointer;
  text-decoration: none;
  opacity: 0.8;
  transition: all 0.5s ease;
}

.btn:hover {
  opacity: 1;
}
