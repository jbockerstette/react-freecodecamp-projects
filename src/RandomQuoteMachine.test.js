import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import RandomQuoteMachine, {
  Grid,
  Quote,
  Button,
  ButtonLink,
  Author
} from './RandomQuoteMachine';

const quotes = [
  {
    quote: 'Hello World',
    author: 'Jim'
  }
];

describe('RandomQuoteMachine', () => {
  const wrapper = shallow(<RandomQuoteMachine />);
  let mountedRandomQuoteMachine;
  let spy;
  fetch.mockResponse(JSON.stringify({ quotes }));
  const app = async () => {
    if (!mountedRandomQuoteMachine) {
      spy = jest.spyOn(RandomQuoteMachine.prototype, 'handleNextQuote');
      mountedRandomQuoteMachine = mount(<RandomQuoteMachine />);
      await mountedRandomQuoteMachine.instance().componentDidMount();
    }
    return mountedRandomQuoteMachine;
  };

  beforeEach(() => {
    mountedRandomQuoteMachine = undefined;
  });

  it('should render a Grid component', () => {
    expect(wrapper.find('Grid').length).toEqual(1);
  });

  it('should render a div #quote-box', () => {
    expect(wrapper.find('#quote-box').length).toEqual(1);
  });

  it('should render a Quote component', () => {
    expect(wrapper.find('Quote').length).toEqual(1);
  });
  it('should render a Author component', () => {
    expect(wrapper.find('Author').length).toEqual(1);
  });
  it('should render a ButtonLink component', () => {
    expect(wrapper.find('ButtonLink').length).toEqual(1);
  });
  it('should render a Button component', () => {
    expect(wrapper.find('Button').length).toEqual(1);
  });

  it('always sets correct state from mock', async () => {
    const app1 = await app();
    const { state } = app1.instance();
    expect(state.quotes.length).toBeGreaterThan(0);
    expect(state.quotes).toMatchObject(quotes);
  });

  it('should have the correct Author', () => {
    const author = shallow(<RandomQuoteMachine />).find('Author');
    expect(author.get(0).props).toHaveProperty('id', 'author');
    expect(author.get(0).props).toHaveProperty('author', '');
  });
  it('should have the correct Quote', () => {
    const quote = shallow(<RandomQuoteMachine />).find('Quote');
    expect(quote.get(0).props).toHaveProperty('id', 'text');
    expect(quote.get(0).props).toHaveProperty('quote', '');
  });
  it('should have the correct ButtonLink', () => {
    const bl = shallow(<RandomQuoteMachine />).find('ButtonLink');
    expect(bl.get(0).props).toHaveProperty('id', 'tweet-quote');
  });
  it('should have the correct Button', () => {
    const btn = shallow(<RandomQuoteMachine />).find('Button');
    expect(btn.get(0).props).toHaveProperty('id', 'new-quote');
    expect(spy).toHaveBeenCalledTimes(0);
    btn.simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
  });
  it('should have the correct quote and author after click', async done => {
    const app1 = await app();
    app1.setState({ quotes: [{ quote: 'hi', author: 'jim' }] }, () => {
      const author = app1.find('Author');
      const quote = app1.find('Quote');
      expect(author.get(0).props).toHaveProperty('author', 'jim');
      expect(quote.get(0).props).toHaveProperty('quote', 'hi');
      done();
    });
    // expect(app1.find('#text').get(0).props).toHaveProperty('opacity', 0);
  });
});

describe('Grid', () => {
  const component = renderer.create(<Grid bgColor="red" />);
  it('always renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Quote', () => {
  const component = renderer.create(
    <Quote id="text" color="red" opacity={0.5} quote="hi" />
  );
  it('always renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Author', () => {
  const component = renderer.create(
    <Author id="author" color="red" opacity={0.5} author="jim" />
  );
  it('always renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {
  const component = renderer.create(
    <Button id="new-quote" bgColor="red" onClick={jest.fn()} />
  );
  it('always renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('ButtonLink', () => {
  const component = renderer.create(
    <ButtonLink
      className="fa fa-twitter"
      id="tweet-quote"
      bgColor="red"
      href="myhref"
      target="_blank"
    />
  );
  it('always renders correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
