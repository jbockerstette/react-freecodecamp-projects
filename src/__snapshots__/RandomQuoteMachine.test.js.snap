// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Author always renders correctly 1`] = `
<p
  id="author"
  style={
    Object {
      "color": "red",
      "opacity": 0.5,
    }
  }
>
  - jim
</p>
`;

exports[`Button always renders correctly 1`] = `
<button
  className="btn"
  id="new-quote"
  onClick={[Function]}
  style={
    Object {
      "background": "red",
    }
  }
/>
`;

exports[`ButtonLink always renders correctly 1`] = `
<a
  className="btn fa fa-twitter"
  href="myhref"
  id="tweet-quote"
  style={
    Object {
      "background": "red",
    }
  }
  target="_blank"
/>
`;

exports[`Grid always renders correctly 1`] = `
<div
  className="grid"
  style={
    Object {
      "background": "red",
    }
  }
/>
`;

exports[`Quote always renders correctly 1`] = `
<p
  id="text"
  style={
    Object {
      "color": "red",
      "opacity": 0.5,
    }
  }
>
  <i
    className="fa fa-quote-left"
  />
   hi 
  <i
    className="fa fa-quote-right"
  />
</p>
`;
